compiler/src/dotty/tools/dotc/typer/Namer.scala

```diff
@@ -1214,8 +1214,16 @@ class Namer { typer: Typer =>
       // it would be erased to BoxedUnit.
       def dealiasIfUnit(tp: Type) = if (tp.isRef(defn.UnitClass)) defn.UnitType else tp
 
-      var rhsCtx = ctx.addMode(Mode.InferringReturnType)
+      var rhsCtx = ctx.fresh.addMode(Mode.InferringReturnType)
       if (sym.isInlineMethod) rhsCtx = rhsCtx.addMode(Mode.InlineableBody)
+      if (typeParams.nonEmpty) {
```

Comment:
Let's refactor this into a common method that is also called from `typedDefDef`.

Reply:
I did not understand why we need to add type parameters to GADT bounds here (or in typedDefDef). This should be documented.

Reply:
Comments were added to both places.